### PR TITLE
Issue 123

### LINKED ISSUE #123: complex numbers with vars
_From @bgoodri on July 18, 2015 15:10_

The following C++ program compiles but segfaults (with clang++-3.6 and g++-5)

```
#include <stan/math.hpp>
#include <iostream>
#include <complex>

int main() {
  stan::math::var x = 1;
  std::complex<stan::math::var> z = x;
  std::cout << "imaginary part is " << z.imag().val() << std::endl;
  return 0;
}
```

I think we should try to prevent that from compiling, or at least dying gracefully, but failing that I guess we should initialize the imaginary part to zero?

_Copied from original issue: stan-dev/stan#1554_

-----

There were a lot of things learned by @ChrisChiasson's effort to implement `std::complex` for Stan's autodiff variables in PR #789. Some of the important takeaways:

- The scope is for both the real and imaginary parts of `std::complex<T>` to be of type `T`. We can either autodiff the real component, the imaginary component, or the result of a function that takes a `std::complex<T>` and returns a `T`. This means we don't have to extend autodiff to deal with adding a `.grad()` function to `std::complex<T>` and we can focus on making sure `std::complex<T>` compiles and behaves appropriately.

- `std::complex<T>` where `T` is our autodiff types (e.g. `stan::math::var`, `stan::math::fvar<U>`) are explicitly unspecified (as of C++11, 08/2018). "The effect of instantiating the template complex for any other type is unspecified." https://en.cppreference.com/w/cpp/numeric/complex

- Given that `std::complex<T>` is unspecified, it is a requirement for any implementation to have tests for the basic operations of `std::complex<T>`. The list of functions can be found here: https://en.cppreference.com/w/cpp/numeric/complex. That means that we should be testing for construction, the member functions, the non-member functions, etc. It's ok to do this in small chunks, but since we're relying on unspecified behavior, we need to be able to tell when our implementation fails in the basic tests before trying to push this through Eigen functions.

- The core reason why `std::complex<T>` can't just be instantiated with one of our autodiff types is because in the current implementation, there are calls to `T()` to construct real or imaginary components. In our current autodiff design, `stan::math::var()` results in an uninitialized variable. This is intentional so we don't create a `vari` on the stack. In order for this to work, we would need to replace `T()` with `T(0)` or some other way of instantiating an autodiff variable on the stack when instantiating a new `std::complex<T>`. For specific places where this happens (especially in the default constructor), search for `_Tp()` in the implementation file: https://github.com/gcc-mirror/gcc/blob/master/libstdc%2B%2B-v3/include/std/complex

- @ChrisChiasson's PR #789 cleverly worked around the above problem. `std::complex<stan::math::var>` was implemented by relying on `stan::math::complex<stan::math::var>` which then relied on `std::complex<stan::zeroing<stan::math::var>>`. `stan::zeroing<stan::math::var>` was a type that had a default constructor that actually made it to the stack. There was a lot of template magic to make this work properly. There is a prototype of this in the closed PR. In order to finish out this design, the full design would need to be written up (it's tricky templating and it extends our meta traits) and fully tested.

- An alternative design might be to write explicit template instantiations for our types for places where it constructs uninitialized autodiff variables for different components. See [this comment](https://github.com/stan-dev/math/pull/789#issuecomment-410898478) for an example of how to template specialize the constructor. If this works, we may only need to specialize a few functions within `std::complex<T>` and not have to use tricky templating. (This is speculation and should be verified with tests.)

- We may need to implement functions in the `std` namespace like `isnan(std::complex<T>)` and `isinf(std::complex<T>)` because Eigen relies on them being in the `std` namespace. @ChrisChiasson was able to do this using ADL by adding these functions to the `stan::math` namespace, but we've already gone ahead and implemented `std::isnan(T)` for both our autodiff types.

### PR DESCRIPTION
**Pull request by bgoodri**
#### Submission Checklist

- [X] Run unit tests: `./runTests.py test/unit`
- [X] Run cpplint: `make cpplint`
- [X] Declare copyright holder and open-source license: see below

#### Summary:

Define some kinda safe `var` constructors when the input is `std::complex`. Eigen uses complex numbers internally for some calculations, so we want to find out if we are accidentally using them.

#### Intended Effect:

For the case where the input is `std::complex` but the imaginary part is zero, create a `var` from the real part. For the case where the input is `std::complex<var>`, just fail.

#### How to Verify:

There are some tests for the non-failing case. I don't know how to test the failing cases because they call `assert`.

#### Side Effects:

Fixes #123 sort of.

R may complain, but at least on Linux the compiler seems to get rid of these constructors when they are not used.

#### Documentation:

None

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Trustees of Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by seantalts** [comment]
It seems like we never use `cassert`'s `assert()` in the math library - do we normally use something else for this kind of error handling? It seems like maybe this could happen at runtime for a user if we allowed complex parameters - should we throw an exception with an error message?

**Comment by bgoodri** [comment]
There is basically no danger of a complex number accidentally being created
by the user. The only danger is that Eigen, Boost, CVODES, etc. use complex
numbers for internal calculations, which would mess up the autodiff at a
minimum. Eigen is known to do this, so I wanted to catch it in the tests.

File: stan/math/rev/core/var.hpp [hunk 1]
```diff
@@ -181,6 +184,56 @@ namespace stan {
...
 
+      /**
+       * Construct a variable from the specified arithmetic argument
+       * by constructing a new <code>vari</code> with the argument
+       * cast to <code>double</code>, and a zero adjoint. Only works
+       * if the imaginary part is zero.
+       *
+       * @param x Value of the variable.
+       */
+      explicit var(std::complex<double> x) {
```
Comment (bob-carpenter) [inline stan/math/rev/core/var.hpp]: Instances of `complex` should be passed by constant reference, `var(const std::complex<double>& x)`.

File: stan/math/rev/core/var.hpp [hunk 2]
```diff
@@ -181,6 +184,56 @@ namespace stan {
...
+        assert(imag(x) == 0);
```
Comment (bob-carpenter) [inline stan/math/rev/core/var.hpp]: This should be an exception.  Although users of the Stan language won't explicitly call it, it's going to be a constructor in the math library. 

I don't know how much printing for complex is built in, but maybe
```
if (imag(x) == 0) {
  std::stringstream ss;
  ss << "Imaginary part of std::complex used to construct var must be zero."
    << " Found argument = " << x << std::endl;
  std::string msg = ss.str();
  throw std::invalid_argument(msg);
}
```

File: stan/math/rev/core/var.hpp [hunk 3]
```diff
@@ -181,6 +184,56 @@ namespace stan {
...
+      /**
+       * Construct a variable from the specified arithmetic argument
+       * by constructing a new <code>vari</code> with the argument
+       * cast to <code>double</code>, and a zero adjoint. Only works
+       * if the imaginary part is zero.
+       *
+       * @param x Value of the variable.
+       */
+      explicit var(std::complex<float> x) {
+        assert(imag(x) == 0);
```
Comment (bob-carpenter) [inline stan/math/rev/core/var.hpp]: same comment as before on pass-by-reference and exception

File: stan/math/rev/core/var.hpp [hunk 4]
```diff
@@ -181,6 +184,56 @@ namespace stan {
...
+      /**
+       * Construct a variable from the specified arithmetic argument
+       * by constructing a new <code>vari</code> with the argument
+       * cast to <code>double</code>, and a zero adjoint. Only works
+       * if the imaginary part is zero.
+       *
+       * @param x Value of the variable.
+       */
+      explicit var(std::complex<long double> x) {
+        assert(imag(x) == 0);
```
Comment (bob-carpenter) [inline stan/math/rev/core/var.hpp]: same comment as previous on exception and pass by reference

File: stan/math/rev/core/var.hpp [hunk 5]
```diff
@@ -181,6 +184,56 @@ namespace stan {
...
+
+      /**
+       * This constructor does not work and really only exists to
+       * ensure that such a var is not created accidentally in
+       * testing.
+       *
+       * @param x Value of the variable.
+       */
+      explicit var(std::complex<var> x) {
+        assert(false);
```
Comment (bob-carpenter) [inline stan/math/rev/core/var.hpp]: I don't think you need this constructor at all.  If you try to call `var(complex<var>(2,0))` it won't have a constructor to match, so the compiler will catch it.  If `complex<T>` can be cast to `T` (I'm not sure, but I'd be surprised), then you want to declare this constructor but not define it.  That will prevent it from being instantiated at compile time.

```
explicit var(const std::complex<var>& x);
```

File: stan/math/rev/core/var.hpp [hunk 6]
```diff
@@ -7,6 +7,9 @@
 #include <boost/math/tools/config.hpp>
 #include <ostream>
 #include <vector>
+#include <complex>
+#include <cassert>
+#include <limits>
```
Comment (bob-carpenter) [inline stan/math/rev/core/var.hpp]: I didn't see anywhere `<limits>` gets used.

You won't need `<cassert>` after changing the asserts to exceptions.

File: stan/math/rev/core/var.hpp (hunk 1 above)
Comment (bgoodri) [reply stan/math/rev/core/var.hpp]: This change results in a compiler error
```
error: cannot bind non-const lvalue reference of type ‘std::complex<long double>&’ to an rvalue of type ‘std::complex<long double>’
```

Comment (bgoodri) [reply stan/math/rev/core/var.hpp]: Nevermind. I forgot the `const`.

File: stan/math/rev/core/var.hpp (hunk 5 above)
Comment (bgoodri) [reply stan/math/rev/core/var.hpp]: This does not prevent the segfault in #123 but neither does filling in the body to only throw an exception. I don't know what to do but am pushing anyway.

Comment (bob-carpenter) [reply stan/math/rev/core/var.hpp]: The segfault is from `complex<var>(var(1))` not `var(complex)`.   

I don't think we can specialize the `complex<T>` constructor to `complex<var>` to fail.

File: stan/math/rev/core/var.hpp [hunk 7]
```diff
@@ -7,6 +7,8 @@
 #include <boost/math/tools/config.hpp>
 #include <ostream>
 #include <vector>
+#include <complex>
+#include <string>
```
Comment (syclik) [inline stan/math/rev/core/var.hpp]: Need to include `stdexcept`.

**Comment by syclik** [comment]
@bob-carpenter, Ben has walked through all the changes. Want to approve? 

I noticed that there's an exception being thrown in the constructor -- we good with that? I don't see us using this yet, but I just want to make sure that's the right pattern to be using for errors.

**Comment by bob-carpenter** [comment]
@syclik I approved it and will let you take over the final review.

I don't see a problem with throwing exceptions in constructors.  Lots of our constructors already do that implicitly for `vari` by calling other functions that can throw exceptions.

Is there something I should be worried about here?

**Comment by syclik** [comment]
Nope. I just wanted to make sure I hadn't considered something obvious with exceptions in constructors.

@bgoodri, add the include and we're good!

**Comment by bob-carpenter** [comment]
I thought this and many of these others were merged.  This is a direct conflict with the cpplint fix for tests and it's going to hit all 15 or so outstanding math pull requests as they changed lots of lines.

**Comment by seantalts** [comment]
@bob-carpenter Git is a lot better at merging than Subversion was - the only outstanding pull request that has a conflict due to the test folder reformatting is mine for removing row vector distribution tests. There are 2 others with unrelated conflicts. I don't think we want to try to coordinate PR merging to try to avoid conflicts - this has been a big waste of time in my professional life whenever it was attempted.

**Comment by seantalts** [comment]
Woops, my bad - #622 and #533 also have conflicts from the test formatting. Still, they should be easy enough to fix and one of those had been sitting there for a while.

**Comment by seantalts** [comment]
Jenkins, retest this please.

**Comment by seantalts** [comment]
Should we merge this?

**Comment by bgoodri** [comment]
I think it does not hurt anything, but it does not help all that much.

**Comment by seantalts** [comment]
Okay, I'll merge it now that the tests are passing. No sense in keeping it around in limbo.

**Comment by ChrisChiasson** [comment]
I am confused as to how this handles issue #123, which is about `std::complex<var>(double)`. From reading the code, this issue handles `var(std::complex<double>)`.

**Comment by bob-carpenter** [comment]
I'm going to close this issue because it's not a concrete technical spec.  First, we need a discussion on the forums, then a functional spec on a wiki, then a technical spec that indicates what needs to be built.  The issue should then get created with the technical spec.

The wrappers will have to be implemented using `complex<var>` for autodiff.  My understanding is that this would let us take derivatives w.r.t. the real component or w.r.t. the complex component, but it wouldn't let us take complex derivatives (derivatives w.r.t. a complex number).

**Comment by bob-carpenter** [comment]
I deleted my last comment after talking to @seantalts and figuring out what's really going on. 

This doesn't directly address #123.  I think the PR is right and the issue was poorly specified here.  

But if the original intent was to just flag that something went wrong, then that's *not* what this PR does---this one just assumes you want to convert a complex number with zero imaginary component to an autodiff variable.

**Pull request merged**